File: 00.scripts/gff2cds.pl
#! /usr/bin/perl
use strict;
use warnings;
use Bio::SeqIO;
use Bio::Seq;

sub translate_nucl{
    my $seq=shift;
    my $seq_obj=Bio::Seq->new(-seq=>$seq,-alphabet=>'dna');
    my $pro=$seq_obj->translate;
    $pro=$pro->seq;
    return($pro);
}

my ($gff,$genome_fa,$cds)=@ARGV;
die "Usage: $0 <gff file> <genome file> <output>\n" if(@ARGV<3);
#my $gff="RefSeq_Btau_4.6.1_protein_coding.gff3";
#my $genome_fa="btau461.fa";
#my $cds="btau461.fa.cds";

my %gff;
my %len;
open(I,"< $gff");
my $no=0;
while(<I>){
    chomp;
    next if(/^#/);
    next if(/^\s*$/);
    my @a=split(/\s+/);
    next unless($a[2] eq "CDS");
    $no++;
    my ($chr,$source,$start,$end,$strand,$phase,$name)=($a[0],$a[1],$a[3],$a[4],$a[6],$a[7],$a[8]);
    # $chr=~s/chr//g;
    $name=~/Parent=([^;]+)/;
    $name=$1;
    if(!exists $len{$name}){
        $len{$name}{chr}=$chr;
        $len{$name}{source}=$source;
        $len{$name}{start}=$start;
        $len{$name}{end}=$end;
        $len{$name}{strand}=$strand;
    }
    else {
        if($start < $len{$name}{start}){
            $len{$name}{start}=$start;
        }
        if($end > $len{$name}{end}){
            $len{$name}{end} = $end;
        }
    }
    $len{$name}{cds}.=$_."\n";
    $gff{$chr}{$name}{$no}{start}=$start;
    $gff{$chr}{$name}{$no}{end}=$end;
    $gff{$chr}{$name}{$no}{strand}=$strand;
    $gff{$chr}{$name}{$no}{phase}=$phase;
}
close I;

my $fa=Bio::SeqIO->new(-file=>$genome_fa,-format=>'fasta');

my %whitelist;
open(O,"> $cds.cds");
open(E,"> $cds.pep");
while(my $seq=$fa->next_seq){
    my $chr=$seq->id;
    my $seq=$seq->seq;
    #print ">$id\n$seq\n";
    next unless(exists $gff{$chr});
    foreach my $name(keys %{$gff{$chr}}){
        my $strand="NA";
        my $line="";
        my $light=0;
        foreach my $no(sort { $gff{$chr}{$name}{$a}{start} <=> $gff{$chr}{$name}{$b}{start} } keys %{$gff{$chr}{$name}}){
            if($strand eq "NA"){
	$strand=$gff{$chr}{$name}{$no}{strand};
            }
            my $start=$gff{$chr}{$name}{$no}{start};
            my $end=$gff{$chr}{$name}{$no}{end};
            my $len=$end-$start+1;
            my $subline=substr($seq,$start-1,$len);
            $line.=$subline;
        }
        if($strand eq "+"){
            my $pep=translate_nucl($line);
            $pep=~/^(.*).$/;
            my $stop_test=$1;
            next if($stop_test=~/\*/);
            $light=1;
            print O ">$name\n$line\n";
            print E ">$name\n$pep\n";
        }
        elsif($strand eq "-"){
            $line=reverse($line);
            $line=~tr/ATCGatcg/TAGCtagc/;
            my $pep=translate_nucl($line);
            $pep=~/^(.*).$/;
            my $stop_test=$1;
            next if($stop_test=~/\*/);
            $light=1;
            print O ">$name\n$line\n";
            print E ">$name\n$pep\n";
        }
        if($light==1){
            $whitelist{$name}=1;
            print STDERR "$chr\t$name\n";
        }
    }
}
close O;
close E;

open O,"> $cds.noStopCodon.gff";
foreach my $name(sort keys %whitelist){
    my $chr=$len{$name}{chr};
    my $source=$len{$name}{source};
    my $start=$len{$name}{start};
    my $end=$len{$name}{end};
    my $strand=$len{$name}{strand};
    my @gene_line=($chr,$source,"gene",$start,$end,".",$strand,".","ID=$name;Parent=$name;");
    my $gene_line=join "\t",@gene_line;
    print O "$gene_line\n";
    my @mrna_line=($chr,$source,"mRNA",$start,$end,".",$strand,".","ID=$name;Parent=$name;");
    my $mrna_line=join "\t",@mrna_line;
    print O "$mrna_line\n";
    print O "$len{$name}{cds}";
}
close O;

# open(I,"< $gff");
# open G,"> $cds.noStopCodon.gff";
# while(<I>){
#     chomp;
#     next if(/^#/);
#     next if(/^\s*$/);
#     my @a=split(/\s+/);
#     next unless($a[2] eq "CDS");
#     my ($chr,$start,$end,$strand,$phase,$name)=($a[0],$a[3],$a[4],$a[6],$a[7],$a[8]);
#     $chr=~s/chr//g;
#     $name=~/Parent=([^;]+)/;
#     $name=$1;
#     next unless(exists $whitelist{$name});
#     print G "$_\n";
# }
# close I;
# close G;

File: 00.scripts/gff2cds.py
#! /usr/bin/env python
import argparse
import re

def read_fasta(fasta_file):
    """读取FASTA文件并返回一个字典，键为序列ID，值为序列字符串。"""
    sequences = {}
    with open(fasta_file, 'r') as f:
        sequence_id = None
        sequence = []
        for line in f:
            line = line.strip()
            if line.startswith(">"):
                if sequence_id:
                    sequences[sequence_id] = ''.join(sequence)
                sequence_id = line[1:].split()[0]  # 取第一个空格前的部分作为ID
                sequence = []
            else:
                sequence.append(line)
        sequences[sequence_id] = ''.join(sequence)  # 添加最后一个序列
    return sequences

def translate_dna_to_protein(dna_seq):
    """将DNA序列翻译成蛋白质序列。这里使用简化的密码子表。"""
    codon_table = {
        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',
        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',
        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',
        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',
        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',
        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',
        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',
        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',
        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',
        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',
        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',
        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',
        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',
        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',
        'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*',
        'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W',
    }
    protein_seq = ''
    # dna_seq = dna_seq.upper()
    for i in range(0, len(dna_seq), 3):
        codon = dna_seq[i:i+3]
        protein_seq += codon_table.get(codon, 'X')  # 使用'X'标记未知的密码子
    return protein_seq

def extract_cds_from_gff(gff_file, genome_sequences):
    """从GFF文件和基因组序列中提取CDS，并按顺序连接每个基因的CDS片段，然后翻译成肽序列。"""
    gene_cds_info = {}
    cds_sequences = {}
    cds_pep_sequences = {}
    gff_lines = {}  # 存储每个CDS ID对应的所有相关GFF行（包括mRNA和CDS）
    mrna_to_cds = {}  # 存储mRNA ID到CDS ID的映射

    with open(gff_file, 'r') as gff:
        for line in gff:
            if line.startswith('#') or not line.strip(): 
                continue
            
            parts = re.split(r'\s+', line.strip())
            if len(parts) < 9: 
                continue

            attributes = parts[8]
            if parts[2] == 'mRNA':
                # 提取mRNA的ID
                mrna_id = None
                for attribute in attributes.split(';'):
                    if attribute.startswith('ID='):
                        mrna_id = attribute.split('=')[1]
                        break
                if mrna_id:
                    if mrna_id not in gff_lines:
                        gff_lines[mrna_id] = []
                    gff_lines[mrna_id].append(line.strip())

            elif parts[2] == 'CDS':
                seq_id = parts[0]
                start = int(parts[3])
                end = int(parts[4])
                strand = parts[6]

                # 提取CDS所属的mRNA ID
                parent_id = None
                for attribute in attributes.split(';'):
                    if attribute.startswith('Parent='):
                        parent_id = attribute.split('=')[1]
                        break

                if not parent_id:
                    print('Error: Parent ID not found for CDS')
                    continue

                # 将CDS行添加到对应mRNA的GFF行列表中
                if parent_id not in gff_lines:
                    gff_lines[parent_id] = []
                gff_lines[parent_id].append(line.strip())

                # 存储CDS信息
                if parent_id not in gene_cds_info:
                    gene_cds_info[parent_id] = []
                gene_cds_info[parent_id].append((seq_id, start, end, strand))

    # 处理每个mRNA的CDS序列
    for mrna_id, cds_parts in gene_cds_info.items():
        cds_parts.sort(key=lambda x: x[1])
        cds_seq_combined = ''
        strand = cds_parts[0][3]  # 获取链的方向

        for seq_id, start, end, _ in cds_parts:
            cds_seq = genome_sequences[seq_id][start-1:end]
            cds_seq_combined += cds_seq

        cds_seq_combined = cds_seq_combined.upper()
        if strand == '-':
            cds_seq_combined = reverse_complement(cds_seq_combined)

        cds_sequences[mrna_id] = cds_seq_combined
        protein_seq = translate_dna_to_protein(cds_seq_combined)
        cds_pep_sequences[mrna_id] = protein_seq

    return cds_sequences, cds_pep_sequences, gff_lines

def reverse_complement(seq):
    """获取DNA序列的反向互补序列。"""
    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}
    return ''.join(complement[base] for base in reversed(seq))

def main(fasta_file, gff_file, output_prefix):
    # 读取基因组序列
    genome_sequences = read_fasta(fasta_file)

    # 提取CDS并翻译成蛋白质序列，同时保留CDS序列和GFF信息
    cds_sequences, cds_pep_sequences, gff_lines = extract_cds_from_gff(gff_file, genome_sequences)

    # 保存CDS序列到文件
    with open(f"{output_prefix}_cds.fasta", 'w') as cds_file:
        for cds_id, cds_seq in cds_sequences.items():
            cds_file.write(f">{cds_id}\n{cds_seq}\n")

    # 保存肽序列到文件，并将GFF信息分别输出到两个文件
    with open(f"{output_prefix}_pep.fasta", 'w') as pep_file, \
         open(f"{output_prefix}_premature_stop_codons.gff", 'w') as stop_file, \
         open(f"{output_prefix}_complete_proteins.gff", 'w') as complete_file:
        
        # 写入GFF文件头部注释
        stop_file.write("##gff-version 3\n")
        complete_file.write("##gff-version 3\n")
        
        for mrna_id, pep_seq in cds_pep_sequences.items():
            if '*' in pep_seq[0:-1]:
                # 将含有提前终止密码子的基因的完整GFF信息（包括mRNA和CDS）写入到停止密码子文件
                for gff_line in gff_lines[mrna_id]:
                    stop_file.write(f"{gff_line}\n")
                continue
            
            # 将成功翻译的基因序列写入到FASTA文件
            pep_file.write(f">{mrna_id}\n{pep_seq}\n")
            
            # 将成功翻译的基因的完整GFF信息（包括mRNA和CDS）写入到完整蛋白质文件
            for gff_line in gff_lines[mrna_id]:
                complete_file.write(f"{gff_line}\n")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Extract CDS and translate to protein sequences from a GFF file and a genome sequence file.")
    parser.add_argument("-g", "--genome_file", required=True, help="Path to the genome sequence file in FASTA format.")
    parser.add_argument("-a", "--annotation_gff_file", required=True, help="Path to the GFF file containing gene annotations.")
    parser.add_argument("-o", "--output_prefix", required=True, help="Prefix for the output files.")

    args = parser.parse_args()

    main(args.genome_file, args.annotation_gff_file, args.output_prefix)